dandiapi/api/views/dandiset.py

```diff
@@ -154,6 +156,54 @@ def get_object(self):
                 raise PermissionDenied()
         return dandiset
 
+    @swagger_auto_schema(
+        query_serializer=DandisetQueryParameterSerializer(),
+        responses={200: DandisetListSerializer(many=True)},
+    )
+    def list(self, request, *args, **kwargs):
+        qs = self.get_queryset().annotate()
+        dandisets = self.paginate_queryset(self.filter_queryset(qs))
+
+        # Query versions, to supply num/count of assets.
+        # Ideally we'd use `distinct('dandiset_id')` to limit the
+        # amount of published versions returned, but we can't use that with `annotate`.
+        versions = (
+            Version.objects.select_related('dandiset')
+            .annotate(
+                total_size=(
+                    Coalesce(Sum('assets__blob__size'), 0)
+                    + Coalesce(Sum('assets__embargoed_blob__size'), 0)
+                    + Coalesce(Sum('assets__zarr__size'), 0)
+                ),
+                num_assets=Count('assets'),
+            )
+            .filter(dandiset__in=dandisets)
+            .order_by('-version', '-modified')
+        )
+
+        # Determine draft and most recently published for each dandiset.
+        #
+        # Since ordering is version descending, then modified descending, we will
+        # see all drafts first, then any published. Once we have seen both one draft
+        # and one published, we're done for that dandiset.
+        dandisets_to_versions = {}
+        for version in versions:
```

Comment:
Both great suggestions, I'll implement them. 



> so we're looking at at least 1000 `Version` objects here, possibly many more since there is no limit to how many versions a dandiset can have.

Yeah this is an unfortunate limitation of the query. I'd realistically like to limit the amount of returned versions to one draft version and one published version (the most recent one) per dandiset. I originally wrote a query that did that, using `distinct` on the dandiset IDs, but annotation/aggregations don't work properly with that. I might add another SQL query to this, since one extra query would probably be worth not iterating over all versions associated with any of the dandisets here.